src/components/MessagesList/MessagesGroup/MessagesSystemGroup.vue

```diff
@@ -0,0 +1,337 @@
+<!--
+  - @copyright Copyright (c) 2023 Maksim Sukharev <antreesy.web@gmail.com>
+  -
+  - @author Maksim Sukharev <antreesy.web@gmail.com>
+  -
+  - @license AGPL-3.0-or-later
+  -
+  - This program is free software: you can redistribute it and/or modify
+  - it under the terms of the GNU Affero General Public License as
+  - published by the Free Software Foundation, either version 3 of the
+  - License, or (at your option) any later version.
+  -
+  - This program is distributed in the hope that it will be useful,
+  - but WITHOUT ANY WARRANTY; without even the implied warranty of
+  - MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+  - GNU Affero General Public License for more details.
+  -
+  - You should have received a copy of the GNU Affero General Public License
+  - along with this program. If not, see <http://www.gnu.org/licenses/>.
+-->
+
+<template>
+	<div class="message-group">
+		<div v-if="dateSeparator"
+			class="message-group__date-header">
+			<span class="date"
+				role="heading"
+				aria-level="3">{{ dateSeparator }}</span>
+		</div>
+		<div class="wrapper wrapper--system">
+			<div v-for="messagesCollapsed of messagesGroupedBySystemMessage"
+				:key="messagesCollapsed.id"
+				class="message-group__system">
+				<ul v-if="messagesCollapsed.messages?.length > 1"
+					class="messages messages--header">
+					<Message v-bind="createCombinedSystemMessage(messagesCollapsed)"
+						is-combined-system-message
+						:is-combined-system-message-collapsed="messagesCollapsed.collapsed"
+						:next-message-id="getNextMessageId(messagesCollapsed.messages.at(-1))"
+						:previous-message-id="getPrevMessageId(messagesCollapsed.messages.at(0))"
+						@toggle-combined-system-message="toggleCollapsed(messagesCollapsed)" />
+				</ul>
+				<ul v-show="messagesCollapsed.messages?.length === 1 || !messagesCollapsed.collapsed"
+					class="messages"
+					:class="{'messages--collapsed': messagesCollapsed.messages?.length > 1}">
+					<Message v-for="message of messagesCollapsed.messages"
+						:key="message.id"
+						v-bind="message"
+						:next-message-id="getNextMessageId(message)"
+						:previous-message-id="getPrevMessageId(message)" />
+				</ul>
+			</div>
```

Comment:
This component has many methods calls in template (that a called during the re-rendering) and those methods are not small utils or formatting methods but data processing methods.

It won't be a problem if it doesn't have unnecessary re-renderings, but currently `MessagesList` re-renders each new message.

For example, a user has a chat with mostly only calls, but large calls, for example, 100 people. Then in the history there will be many groups with 100 system messages about leaving the call. And every message, reaction, poll update will re-render the list with recalculating all these methods looping over 100 messages for each group.

IMO, we can keep it in this way but then do one of:
- Find a reason of re-renderings
- Refactor all these components, remove unnecessary props
- Remove these calculating inside rendering
- Wait for Vue 3 migration =D

Reply:
Also, I'd prefer to use `in` in `v-for` directives to use only one style. (`v-for of` is not very often used in our components).

Reply:
Moved to follow-up